design/proposals/module-linking/Binary.md

```diff
@@ -1,198 +1,174 @@
 ```
-aliassec ::=  a*:section_16(vec(alias))     ->        a*
-
-alias ::=
-    0x00 i:instanceidx 0x00 nm:name         ->        (alias $i "nm" (func))
-    0x00 i:instanceidx 0x01 nm:name         ->        (alias $i "nm" (table))
-    0x00 i:instanceidx 0x02 nm:name         ->        (alias $i "nm" (memory))
-    0x00 i:instanceidx 0x03 nm:name         ->        (alias $i "nm" (global))
-    0x00 i:instanceidx 0x05 nm:name         ->        (alias $i "nm" (module))
-    0x00 i:instanceidx 0x06 nm:name         ->        (alias $i "nm" (instance))
-    0x01 ct:varu32 0x05 m:moduleidx         ->        (alias outer ct m (module))
-    0x01 ct:varu32 0x07 t:typeidx           ->        (alias outer ct m (type))
+alias    ::= 0x00 i:<instanceidx> nm:<name> d:<def-kind> => (alias i nm d)
+           | 0x01 ct:<varu32> li:<localidx> d:<def-kind> => (alias outer ct li d)
+def-kind ::= 0x00                                        => (instance)
+           | 0x01                                        => (module)
+           | 0x02                                        => (func)
+           | 0x03                                        => (type)
+           | 0x10                                        => (table)
+           | 0x11                                        => (memory)
+           | 0x12                                        => (global)
 ```
+Notes:
+* For export aliases, `i` is validated to refer to an instance in the instance
+  index space that actually exports `nm` with the specified `def-kind`.
+* For outer aliases, `ct` is validated to be *less than* the number of enclosing
+  modules and `li` is validated to be a valid index in the `def-kind` index space.
+* For the time being, export aliases may not have the `type` `def-kind` and outer
+  aliases may *only* have the `module` or `type` `def-kind`s.
 
-**Validation**
-
-* The instance, module and type indices are all in bounds. Remember that "in bounds"
-  means only those definitions preceding this alias definition in binary format
-  order. In the case of outer aliases, this means the position of the nested module
-  definition in the outer module.
-* Aliased instance export names must be found in `i`'s instance type with a
-  matching definition kind.
-* The `ct` of an outer alias must be *less than* the number of enclosing modules
-  which implicitly prevents outer aliases from appearing in top-level modules.
-
-## Function section
-
-**Validation**
-
-* Type indices must point to function types.
 
-## Export section
-
-update
-[`exportdesc`](https://webassembly.github.io/spec/core/binary/modules.html#binary-exportdesc)
+## Type Definitions
 
+(See [type definitions](Explainer.md#type-definitions) in the explainer.)
 ```
-exportdesc ::=
-    ...
-    0x05 x:moduleidx                        -> module x
-    0x06 x:instanceidx                      -> instance x
+type             ::= 0x7f e*:vec(export-decl)               -> (instance e*)
```

Comment:
Instance types need to be able to include type definitions and aliases as well, otherwise they can neither define interesting type nor refer to outer definitions. Likewise, some dependent signatures wouldn't be expressible in the presence of type exports. For example:
```
(instance
  (instance $I (export "I") (type (export "T")))
  (type $T (alias $I "T"))
  (type $F (func))
  (func (export "f") (param (ref $T) (ref $F)))
)
```
So they have to be a sublanguage of module types only excluding imports.

Reply:
x2